File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/crucible/CrucibleHeatCategory.java
package wraith.fabricaeexnihilo.compatibility.rei.crucible;

import me.shedaniel.math.Point;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.Renderer;
import me.shedaniel.rei.api.client.gui.widgets.Widget;
import me.shedaniel.rei.api.client.gui.widgets.Widgets;
import me.shedaniel.rei.api.client.registry.display.DisplayCategory;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.text.TranslatableText;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;

import java.util.ArrayList;
import java.util.List;

public class CrucibleHeatCategory implements DisplayCategory<CrucibleHeatDisplay> {

    private final ItemStack icon;
    private final String name;

    public CrucibleHeatCategory(ItemStack icon, String name) {
        this.icon = icon;
        this.name = name;
    }

    @Override
    public CategoryIdentifier<? extends CrucibleHeatDisplay> getCategoryIdentifier() {
        return PluginEntry.HEATING;
    }

    @Override
    public Renderer getIcon() {
        return EntryStacks.of(icon);
    }

    @Override
    public Text getTitle() {
        return new TranslatableText(this.name);
    }

    @Override
    public int getDisplayHeight() {
        return 49;
    }

    @Override
    public List<Widget> setupDisplay(CrucibleHeatDisplay display, Rectangle bounds) {
        var widgets = new ArrayList<Widget>();
        widgets.add(Widgets.createRecipeBase(bounds));

        var inputs = display.getInputEntries().get(0);
        var heat = display.getHeat();

        Point startPoint = new Point(bounds.getCenterX() - 41, bounds.getCenterY() - 17);
        widgets.add(
            Widgets.createLabel(new Point(bounds.x + 26, bounds.getCenterY()), new TranslatableText("fabricaeexnihilo.rei.category.crucible_heat.speed", heat))
                .color(0xFF404040, 0xFFBBBBBB)
                .noShadow()
                .leftAligned()
        );
        widgets.add(Widgets.createBurningFire(new Point(bounds.x + 6, startPoint.y + 1)).animationDurationTicks(100F / heat));
        widgets.add(Widgets.createSlot(new Point(bounds.x + 6, startPoint.y + 18)).entries(inputs).markInput());
        return widgets;
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/crucible/CrucibleHeatCategory.kt
package exnihilofabrico.compatibility.rei.crucible

import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.util.asREIEntry
import exnihilofabrico.util.getExNihiloItemStack
import me.shedaniel.math.api.Point
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.api.RecipeCategory
import me.shedaniel.rei.gui.widget.EntryWidget
import me.shedaniel.rei.gui.widget.LabelWidget
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import me.shedaniel.rei.gui.widget.Widget
import java.util.function.Supplier

class CrucibleHeatCategory(): RecipeCategory<CrucibleHeatDisplay> {

    override fun getIdentifier() = PluginEntry.CRUCIBLE_HEAT
    override fun getLogo() = getExNihiloItemStack("stone_crucible").asREIEntry()
    override fun getCategoryName() = "Heat"


    override fun getDisplayHeight() = HEIGHT
    override fun getDisplayWidth(display: CrucibleHeatDisplay) =
        WIDTH

    override fun setupDisplay(displaySupplier: Supplier<CrucibleHeatDisplay>, bounds: Rectangle): MutableList<Widget> {
        val display = displaySupplier.get()
        val widgets = mutableListOf<Widget>(RecipeBaseWidget(bounds))

        val inputs = display.inputEntries[0]
        val heat = display.getHeat()

        // Block
        widgets.add(EntryWidget.create(bounds.minX + BLOCK_X, bounds.minY + BLOCK_Y).entries(inputs))
        // Heat Value
        val text =  LabelWidget(0, 0, "${heat}")
        text.position = Point(bounds.minX + WIDTH - MARGIN - text.bounds.maxX, bounds.minY + MARGIN + 18 - text.bounds.maxY - text.bounds.maxY / 2)

        widgets.add(text)

        return widgets
    }

    companion object {

        val MARGIN = 6
        val WIDTH = MARGIN*2 + 18*4
        val HEIGHT = MARGIN*2 + 16

        val BLOCK_X = MARGIN
        val BLOCK_Y = MARGIN
    }

}